src/event_notifier.rs

```diff
@@ -0,0 +1,291 @@
+    /// #[cfg(any(target_os = "linux", target_os = "android"))]
+    /// use vmm_sys_util::event_notifier::EFD_NONBLOCK;
+    /// #[cfg(target_os = "macos")]
+    /// use vmm_sys_util::event_notifier::O_NONBLOCK;
+    ///
+    /// #[cfg(any(target_os = "linux", target_os = "android"))]
+    /// let evt = EventNotifier::new(EFD_NONBLOCK).unwrap();
+    /// #[cfg(target_os = "macos")]
+    /// let evt = EventNotifier::new(O_NONBLOCK).unwrap();
+    /// evt.write(55).unwrap();
+    /// assert_eq!(evt.read().unwrap(), 55);
+    /// ```
+    pub fn read(&self) -> result::Result<u64, io::Error> {
+        let mut buf: u64 = 0;
+        // SAFETY: This is safe because we made this fd and the pointer we
+        // pass can not overflow because we give the syscall's size parameter properly.
+        let ret = unsafe {
+            read(
+                self.rfd.as_raw_fd(),
+                &mut buf as *mut u64 as *mut c_void,
+                mem::size_of::<u64>(),
+            )
+        };
+        if ret < 0 {
+            Err(io::Error::last_os_error())
+        } else {
+            Ok(buf)
+        }
+    }
+
+    /// Clone this EventFd or PipeFd.
+    ///
+    /// This internally creates a new file descriptor and it will share the same
+    /// underlying count within the kernel.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// extern crate vmm_sys_util;
+    /// use vmm_sys_util::event_notifier::EventNotifier;
+    /// #[cfg(any(target_os = "linux", target_os = "android"))]
+    /// use vmm_sys_util::event_notifier::EFD_NONBLOCK;
+    /// #[cfg(target_os = "macos")]
+    /// use vmm_sys_util::event_notifier::O_NONBLOCK;
+    /// #[cfg(any(target_os = "linux", target_os = "android"))]
+    /// let evt = EventNotifier::new(EFD_NONBLOCK).unwrap();
+    /// #[cfg(target_os = "macos")]
+    /// let evt = EventNotifier::new(O_NONBLOCK).unwrap();
+    /// let evt_clone = evt.try_clone().unwrap();
+    /// evt.write(923).unwrap();
+    /// assert_eq!(evt_clone.read().unwrap(), 923);
+    /// ```
+    pub fn try_clone(&self) -> result::Result<EventNotifier, io::Error> {
+        // SAFETY: This is safe because we made this fd and properly check that it returns
+        // without error.
+        let (rfd, wfd) = unsafe { (dup(self.rfd.as_raw_fd()), dup(self.wfd.as_raw_fd())) };
+        if rfd < 0 || wfd < 0 {
+            Err(io::Error::last_os_error())
+        } else {
+            Ok(EventNotifier {
+                // SAFETY: This is safe because we checked ret for success and know the kernel
+                // gave us an fd that we own.
+                rfd: unsafe { File::from_raw_fd(rfd) },
+                wfd: unsafe { File::from_raw_fd(wfd) },
+            })
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_new() {
+        #[cfg(any(target_os = "linux", target_os = "android"))]
+        EventNotifier::new(EFD_NONBLOCK).unwrap();
```

Comment:
Maybe we'll want to create an enum containing the flags that are common between Mac and linux, which then internally get translated into the platform-specific flags in some sort of utility constructor, so that callers that want to be agnostic don't have to do with cfg-ery? (but keep a version of the constructor that takes a raw `flags` argument, as we dont want to restrict usage of this library to flags that are common between mac and linux)